Implement `s.split()` for `Stream + Sink`

tailhook opened the pull request:
Basically, this replaces `tokio_core::io::Framed::split`, it's unclear why framed needs specific implementation instead of generic one like this.

Also, I think we should remove `Stream` and `Sink` for `BiLock<T>` in favor of this idea. I don't think using Stream on both instances of the same BiLock is generally useful.

alexcrichton commented on src/stream/split.rs:
```diff
@@ -0,0 +1,39 @@
+        let (a, b) = BiLock::new(s);
+        let read = SplitStream(a);
+        let write = SplitSink(b);
+        (read, write)
```
indentation here may be a bit off?

alexcrichton commented:
Yeah let's go a head and remove the impls directly on `BiLock` itself, this makes more sense to me 👍

alexcrichton commented:
Also we tend to prefer `let (tx, rx) = foo()` for pair-returning methods, so perhaps the order of the return value could be swapped?

tailhook commented:
I have to admit that we use `(tx, rx)` for things returning channel. This use case is quite opposite. However, it may happen that `split()` is useful for channel-like things too (not very sure though), and it's still easier to remember if order is same. So, I've changed method to return `(sink, stream)`.

Also fixed indent, rebased and removed Stream/Sink for BiLock.

tailhook commented:
Okay, last amendment fixes `--no-default-features`

alexcrichton commented:
Thanks!
